sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java

```diff
@@ -0,0 +1,1071 @@
+     */
+    public <InputT, OutputT> ComposedCombineFn<DataT> with(
+        SimpleFunction<DataT, InputT> extractInputFn,
+        CombineFn<InputT, ?, OutputT> combineFn,
+        TupleTag<OutputT> outputTag) {
+      checkUniqueness(outputTags, outputTag);
+      return new ComposedCombineFn<>(
+          ImmutableList.<SerializableFunction<DataT, ?>>builder()
+              .addAll(extractInputFns)
+              .add(extractInputFn)
+              .build(),
+          ImmutableList.<CombineFn<?, ?, ?>>builder()
+              .addAll(combineFns)
+              .add(combineFn)
+              .build(),
+          ImmutableList.<TupleTag<?>>builder()
+              .addAll(outputTags)
+              .add(outputTag)
+              .build());
+    }
+
+    /**
+     * Returns a {@link ComposedCombineFnWithContext} with an additional
+     * {@link CombineFnWithContext}.
+     */
+    public <InputT, OutputT> ComposedCombineFnWithContext<DataT> with(
+        SimpleFunction<DataT, InputT> extractInputFn,
+        CombineFnWithContext<InputT, ?, OutputT> combineFn,
+        TupleTag<OutputT> outputTag) {
+      checkUniqueness(outputTags, outputTag);
+      List<CombineFnWithContext<Object, Object, Object>> fnsWithContext = Lists.newArrayList();
+      for (CombineFn<Object, Object, Object> fn : combineFns) {
+        fnsWithContext.add(toFnWithContext(fn));
+      }
+      return new ComposedCombineFnWithContext<>(
+          ImmutableList.<SerializableFunction<DataT, ?>>builder()
+              .addAll(extractInputFns)
+              .add(extractInputFn)
+              .build(),
+          ImmutableList.<CombineFnWithContext<?, ?, ?>>builder()
+              .addAll(fnsWithContext)
+              .add(combineFn)
+              .build(),
+          ImmutableList.<TupleTag<?>>builder()
+              .addAll(outputTags)
+              .add(outputTag)
+              .build());
+    }
+
+    @Override
+    public Object[] createAccumulator() {
+      Object[] accumsArray = new Object[combineFnCount];
+      for (int i = 0; i < combineFnCount; ++i) {
+        accumsArray[i] = combineFns.get(i).createAccumulator();
+      }
+      return accumsArray;
+    }
+
+    @Override
+    public Object[] addInput(Object[] accumulator, DataT value) {
+      for (int i = 0; i < combineFnCount; ++i) {
+        Object input = extractInputFns.get(i).apply(value);
+        accumulator[i] = combineFns.get(i).addInput(accumulator[i], input);
+      }
+      return accumulator;
+    }
+
+    @Override
+    public Object[] mergeAccumulators(Iterable<Object[]> accumulators) {
+      Iterator<Object[]> iter = accumulators.iterator();
+      if (!iter.hasNext()) {
+        return createAccumulator();
+      } else {
+        Object[] accum = iter.next();
+        for (int i = 0; i < combineFnCount; ++i) {
+          accum[i] = combineFns.get(i).mergeAccumulators(new ProjectionIterable(accumulators, i));
```

Comment:
I assume that ProjectIterable(accumulators, i) will skip the first item in accumulators since we've already pulled that out? If yes, please document. If no, that seems wrong.